### ISSUE TITLE
(#feat): Pointcut allows patterns to choose methods

### ISSUE THREAD

**Issue by konstantinkrassmann** [amoilanen/jsAspect#14]
Hello,
Inspired by jAspect it should be possible to define a regExp to select, what methods should be intercepted. 

E.g. I only want to intercept all methods starting with "get*":

``` javascript
var afterAdvice =  new jsAspect.Advice.After("get*", function() {
  console.log("joinPoint", "after");
});
```

`getName` should match and `setName` shouldn't. 

This is very usefull, because currently you have to do this method-by-method or for all methods at the same time..

**Comment by amoilanen** [comment]
Nice feature, to be implemented, this indeed adds more flexibility to how pointcuts can be defined.

**Comment by amoilanen** [comment]
Implemented in 367f5c8a35...9d07c4873a

However, the regular expression is not defined per advice. The advice is just some wrapped function that knows how to be applied at a certain point when other functions are executed. The right place to specify a regular expression seems to be in the aspect.

Examples of the new API can be found in the added tests, for example:

``` javascript
    var aspect = new jsAspect.Aspect([
      new jsAspect.Advice.After(function(context) {
        console.log(context.method.name);
      })
    ]);

    aspect.withRegex("get.*").applyTo(Target);
```

The regular expressions used are the JavaScript regular expressions so rather than writing `get*` we should write `get.*`. This way we get all the power of the standard engine of regular expressions and the syntax is also familiar.

Please, let me know if you have any further comments or suggestions.

**Comment by konstantinkrassmann** [comment]
Good changes there !

In jAspect and in the definition of aop, such selectors eg. by regexp are basically the pointcuts. 

The definition is, that pointcuts define the set of functions, that need to be intercepted to advice them.
What we yet define as pointcuts are actually scopes. 
(java doesn't have the seperation of prototype and simple object functions, so there's no need of scopes)

Maybe rename the current pointcuts to scopes and define those regexps as real pointcuts.

Applying the Regex on the aspect is nice to have, but i d love to see it on advices. 

For example:
I want to define 2 different advices to the __before_ join point. But the first advice should be applied to my `get.*` methods and the other should be applied to my `set*` method. 

And i just want the advice __afterReturning_ only on my getters (cuz setters doesnt return anything, waste of logs).

``` javascript
var aspect = new jsAspect.Aspect([
      new jsAspect.Advice.Before("get.*",function(context) {
        console.log("getting", context.method.arguments);
      }),
     new jsAspect.Advice.Before("set.*",function(context) {
        console.log("setting", context.method.arguments);
      }),
     new jsAspect.Advice.AfterReturning("get.*",function(context) {
        if(content.returnValue == null){
            throw "Null returned";
        }
      })
]);

aspect.applyTo(Target);
```

**Comment by amoilanen** [comment]
Makes sense. Implemented in 9d07c4873a...0583bd0c6c, the API is slightly different which in my opinion looks more readable. For example, we, probably, should not put the regular expression at the first place as it then looks like a mandatory argument:

``` javascript
var aspect = new jsAspect.Aspect([
  new jsAspect.Advice.Before(function(context) {
    console.log("getting", context.method.arguments);
  }).withRegex("get.*"),
  new jsAspect.Advice.Before(function(context) {
    console.log("setting", context.method.arguments);
  }).withRegex("set.*"),
  new jsAspect.Advice.AfterReturning(function(context) {
    if(content.returnValue == null){
      throw "Null returned";
    }
  }).withRegex("get.*")
]);

aspect.applyTo(Target);
```

**Comment by konstantinkrassmann** [comment]
Good job, that's what i wanted!
